File: supabase/functions/chatbot-whatsapp-webhook/chatbot/conversation_state.ts
import { createClient } from "https://esm.sh/@supabase/supabase-js@2"

const supabase = createClient(
  Deno.env.get('SUPABASE_URL') ?? '',
  Deno.env.get('SUPABASE_SERVICE_ROLE_KEY') ?? ''
)

export interface ConversationState {
  step: 'main_menu' | 'make_order' | 'new_order_warning' | 'order_type' | 'address_selection' | 'new_address' | 'store_selection' | 'store_search' | 'menu_display' | 'category_items' | 'item_details' | 'item_customization' | 'cart_view' | 'checkout' | 'payment' | 'tip_selection' | 'email_collection' | 'payment_method' | 'review_selection' | 'rating' | 'comment' | 'comment_write' | 'anon_selection' | 'completed' | 'address_list' | 'address_actions' | 'address_add' | 'address_added' | 'location_suggestions' | 'profile_view' | 'profile_edit' | 'profile_field_edit'
  
  // Order flow data
  delivery_type?: 'delivery' | 'collection'
  selected_address_id?: string
  selected_location_id?: string
  current_category_id?: string
  current_item_id?: string
  
  // Menu pagination data
  menu_page?: number
  menu_view_type?: 'categories' | 'items'
  
  // Item customization data
  temp_item_variations?: Record<string, string> // variation_id -> option_id
  current_variation_index?: number
  
  // Cart data
  cart_items?: Array<{
    item_id: string
    quantity: number
    variations: Record<string, string>
    special_instructions?: string
  }>
  
  // Payment data
  tip_amount?: number
  customer_email?: string
  payment_method?: string
  payment_authorization_id?: string
  
  // Review data (existing)
  selected_bite_id?: string
  rating?: number
  comment?: string
  review_page?: number
  
  // Profile data
  editing_field?: 'first_name' | 'last_name' | 'email'
  
  // Navigation
  previous_step?: string
  temp_address_data?: any
}

export async function getConversationState(chatId: string): Promise<ConversationState> {
  const { data: chat, error } = await supabase
    .from('chats')
    .select('conversation_state')
    .eq('id', chatId)
    .single()
    
  if (error) {
    console.error('Error getting conversation state:', error)
    return { step: 'main_menu' }
  }
  
  return chat.conversation_state || { step: 'main_menu' }
}

export async function updateConversationState(chatId: string, newState: ConversationState): Promise<void> {
  const { error } = await supabase
    .from('chats')
    .update({ 
      conversation_state: newState,
      updated_at: new Date().toISOString()
    })
    .eq('id', chatId)
    
  if (error) {
    console.error('Error updating conversation state:', error)
  }
}

export async function resetConversationState(chatId: string): Promise<void> {
  await updateConversationState(chatId, { step: 'main_menu' })
} 

File: backend/internal/chatbot/conversation_state.go
package chatbot

import (
	"context"
	"encoding/json"
	"log"
)

// ConversationState mirrors the TS ConversationState interface.
// Pointer fields are used for nullable primitives so we can distinguish
// "unset" from the zero value — matching JS `undefined` behavior.
type ConversationState struct {
	Step string `json:"step"`

	// Order flow data
	DeliveryType       string `json:"delivery_type,omitempty"`
	SelectedAddressID  string `json:"selected_address_id,omitempty"`
	SelectedLocationID string `json:"selected_location_id,omitempty"`
	CurrentCategoryID  string `json:"current_category_id,omitempty"`
	CurrentItemID      string `json:"current_item_id,omitempty"`

	// Menu pagination data
	MenuPage     int    `json:"menu_page,omitempty"`
	MenuViewType string `json:"menu_view_type,omitempty"`

	// Item customization data
	TempItemVariations    map[string]string `json:"temp_item_variations,omitempty"`
	CurrentVariationIndex *int              `json:"current_variation_index,omitempty"`

	// Payment data
	TipAmount              *float64 `json:"tip_amount,omitempty"`
	CustomerEmail          string   `json:"customer_email,omitempty"`
	PaymentMethod          string   `json:"payment_method,omitempty"`
	PaymentAuthorizationID string   `json:"payment_authorization_id,omitempty"`

	// Review data
	SelectedBiteID string `json:"selected_bite_id,omitempty"`
	Rating         *int   `json:"rating,omitempty"`
	Comment        string `json:"comment,omitempty"`
	ReviewPage     int    `json:"review_page,omitempty"`

	// Profile data
	EditingField string `json:"editing_field,omitempty"`

	// Billing data
	SelectedPaymentMethodID string `json:"selected_payment_method_id,omitempty"`

	// Navigation
	PreviousStep    string      `json:"previous_step,omitempty"`
	TempAddressData *TempAddress `json:"temp_address_data,omitempty"`
}

// TempAddress matches the loose TS temp_address_data shape.
type TempAddress struct {
	AddressLine1 string       `json:"address_line_1,omitempty"`
	Address      string       `json:"address,omitempty"`
	Coordinates  *Coordinates `json:"coordinates,omitempty"`
	Suggestions  []string     `json:"suggestions,omitempty"`
}

type Coordinates struct {
	Latitude  float64 `json:"latitude"`
	Longitude float64 `json:"longitude"`
}

func (s *Service) getConversationState(ctx context.Context, chatID string) ConversationState {
	var raw []byte
	err := s.pool.QueryRow(ctx,
		`SELECT conversation_state FROM chats WHERE id = $1`,
		chatID,
	).Scan(&raw)
	if err != nil {
		log.Printf("Error getting conversation state: %v", err)
		return ConversationState{Step: "main_menu"}
	}
	if len(raw) == 0 {
		return ConversationState{Step: "main_menu"}
	}
	var state ConversationState
	if err := json.Unmarshal(raw, &state); err != nil {
		log.Printf("Error decoding conversation state: %v", err)
		return ConversationState{Step: "main_menu"}
	}
	if state.Step == "" {
		state.Step = "main_menu"
	}
	return state
}

func (s *Service) updateConversationState(ctx context.Context, chatID string, newState ConversationState) {
	raw, err := json.Marshal(newState)
	if err != nil {
		log.Printf("Error encoding conversation state: %v", err)
		return
	}
	_, err = s.pool.Exec(ctx,
		`UPDATE chats SET conversation_state = $1::jsonb, updated_at = NOW() WHERE id = $2`,
		string(raw), chatID,
	)
	if err != nil {
		log.Printf("Error updating conversation state: %v", err)
	}
}

func (s *Service) resetConversationState(ctx context.Context, chatID string) {
	s.updateConversationState(ctx, chatID, ConversationState{Step: "main_menu"})
}

// intPtr and floatPtr are small helpers used across handlers.
func intPtr(v int) *int          { return &v }
func floatPtr(v float64) *float64 { return &v }
